File: business-svc/src/main/java/org/zanata/magpie/service/DocumentContentTranslatorService.java
package org.zanata.magpie.service;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;
import java.util.stream.Collectors;
import javax.enterprise.context.ApplicationScoped;
import javax.inject.Inject;
import javax.ws.rs.BadRequestException;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;

import org.apache.commons.lang3.StringUtils;
import org.jsoup.nodes.Element;
import org.jsoup.nodes.Node;
import org.jsoup.nodes.TextNode;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.zanata.magpie.api.dto.APIResponse;
import org.zanata.magpie.api.dto.DocumentContent;
import org.zanata.magpie.api.dto.TypeString;
import org.zanata.magpie.model.Document;
import org.zanata.magpie.model.StringType;
import org.zanata.magpie.util.ArticleUtil;
import org.zanata.magpie.exception.MTException;
import org.zanata.magpie.model.BackendID;

import com.google.common.collect.ImmutableList;
import org.zanata.magpie.util.ShortString;
import static org.zanata.magpie.util.SegmentStringKt.segmentBySentences;

/**
 *
 * @author Alex Eng <a href="mailto:aeng@redhat.com">aeng@redhat.com</a>
 */
@ApplicationScoped
public class DocumentContentTranslatorService {
    private static final Logger LOG =
            LoggerFactory.getLogger(DocumentContentTranslatorService.class);

    // set MT category for translation
    private static final String CATEGORY = "tech";

    private PersistentTranslationService persistentTranslationService;

    @SuppressWarnings({"unused", "uninitialized"})
    public DocumentContentTranslatorService() {
    }

    @Inject
    public DocumentContentTranslatorService(
            PersistentTranslationService persistentTranslationService) {
        this.persistentTranslationService = persistentTranslationService;
    }

    /**
     * Translate a Document and send in for machine translation request.
     *
     * If string is no longer than maxLength, it will be send as a batch according
     * to the type.
     *
     * If a string is longer than maxLength,
     * Plain Text: try to segment it and send it as a single MT request. Will
     * be ignored if cannot be segmented.
     * HTML: try to run through the node tree and translate. Will be ignored if
     * cannot be broken down.
     *
     * {@link DocumentContent}
     **/
    public DocumentContent translateDocument(Document doc,
            DocumentContent documentContent, BackendID backendID)
            throws BadRequestException, MTException {
        Map<Integer, TypeString> indexTextMap = new LinkedHashMap<>();
        Map<Integer, TypeString> indexHTMLMap = new LinkedHashMap<>();
        Map<Integer, TypeString> indexXMLMap = new LinkedHashMap<>();
        int maxLength = persistentTranslationService.getMaxLength(backendID);

        List<APIResponse> warnings = new ArrayList<>();
        List<TypeString> typeStrings =
                new ArrayList<>(documentContent.getContents());

        // counts short strings (less than maxLength) which are translated
        // in a batch at the end
        int index = 0;
        for (TypeString typeString: typeStrings) {
            MediaType mediaType = getMediaType(typeString.getType());
            String source = typeString.getValue();
            if (mediaType.equals(MediaType.TEXT_PLAIN_TYPE)
                    && !StringUtils.isBlank(source)) {
                if (source.length() <= maxLength) {
                    indexTextMap.put(index, typeString);
                } else {
                    StringTranslationResult result =
                            translatePlainTextBySentences(doc, backendID,
                                    source, maxLength);
                    typeString.setValue(result.getTranslation());
                    warnings.addAll(result.getWarnings());
                }
            } else if (mediaType.equals(MediaType.TEXT_HTML_TYPE) ||
                    mediaType.equals(MediaType.TEXT_XML_TYPE)) {
                boolean xml = mediaType.equals(MediaType.TEXT_XML_TYPE);

                // put placeholders in all non-translatable elements
                TranslatableNodeList translatableNodeList = xml ?
                        ArticleUtil.replaceNonTranslatableNodeXML(index, source) :
                        ArticleUtil.replaceNonTranslatableNodeHTML(index, source);
                String html = translatableNodeList.getHtml();
                if (html.length() <= maxLength) {
                    Map<Integer, TypeString> indexMap = xml ? indexXMLMap : indexHTMLMap;
                    indexMap.put(index, typeString);
                } else {
                    Function<String, Element> toElement = xml ? ArticleUtil::wrapXML : ArticleUtil::wrapHTML;
                    StringTranslationResult result =
                            translateLargeElement(doc, backendID, mediaType,
                                    maxLength, html, toElement);
                    String translationWithPlaceholders = result.getTranslation();

                    // replace placeholder with original node
                    String translatedString =
                            removePlaceholders(translationWithPlaceholders,
                                    translatableNodeList, toElement);
                    typeString.setValue(translatedString);
                    warnings.addAll(result.getWarnings());
                }
            }
            index++;
        }

        translateAndMergeStringsInBatch(doc, backendID, StringType.TEXT_PLAIN,
                maxLength, indexTextMap, typeStrings);
        translateAndMergeStringsInBatch(doc, backendID, StringType.HTML,
                maxLength, indexHTMLMap, typeStrings);
        translateAndMergeStringsInBatch(doc, backendID, StringType.XML,
                maxLength, indexXMLMap, typeStrings);
        return new DocumentContent(typeStrings, documentContent.getUrl(),
                doc.getToLocale().getLocaleCode().getId(), backendID.getId(),
                warnings);
    }

    private String removePlaceholders(String translationWithPlaceholders,
            TranslatableNodeList translatableNodeList,
            Function<String, Element> toElement) {
        return ArticleUtil.replacePlaceholderWithNode(
                translatableNodeList.getPlaceholderIdMap(),
                translationWithPlaceholders, toElement);
    }

    /**
     * Translate strings with batch of maxLength
     */
    private StringTranslationResult translatePlainTextBySentences(Document doc,
            BackendID backendID, String sourceText, int maxBatchLength) {
        List<APIResponse> warnings = new ArrayList<>();
        List<String> sourceSentences =
                segmentBySentences(sourceText,
                        Optional.of(doc.getFromLocale().getLocaleCode()));
        // source sentences which have been collected in a batch
        List<String> batchSentences = new ArrayList<>();
        // invariant: should equal total number of chars in batchSentences
        int charsInBatch = 0;
        // the indices (within sourceSentences) of sentences short enough to translate
        List<Integer> translatableSentenceNums = new ArrayList<>();
        // the translations of the translatable sentences. Same size as translatableSentenceNums?
        List<String> translatedSentences = new ArrayList<>();
        for (int sourceSentenceNum = 0; sourceSentenceNum < sourceSentences.size(); sourceSentenceNum++) {
            String sourceSentence = sourceSentences.get(sourceSentenceNum);
            // ignore string if length is longer than maxLength
            if (sourceSentence.length() > maxBatchLength) {
                warnings.add(maxLengthWarning(sourceSentence, maxBatchLength));
                continue;
            }
            if (charsInBatch + sourceSentence.length() > maxBatchLength) {
                // Adding this sentence to the batch would take us over the
                // limit, so process the previous batch now.
                processBatchSentences(doc, backendID, batchSentences,
                        translatedSentences);
                charsInBatch = 0;
            }
            // add sentence to the batch (which may be brand new)
            batchSentences.add(sourceSentence);
            charsInBatch += sourceSentence.length();
            translatableSentenceNums.add(sourceSentenceNum);
        }
        if (!batchSentences.isEmpty()) {
            // translate the leftovers in a last batch
            processBatchSentences(doc, backendID, batchSentences, translatedSentences);
            // just maintaining the invariant (for completeness):
            //noinspection UnusedAssignment
            charsInBatch = 0;
        }

        List<String> results = new ArrayList<>(sourceSentences);
        for (int index = 0; index < translatedSentences.size(); index++) {
            results.set(translatableSentenceNums.get(index), translatedSentences.get(index));
        }
        return new StringTranslationResult(String.join("", results), warnings);
    }

    private void processBatchSentences(Document doc, BackendID backendID,
            List<String> batchSentences, List<String> translatedSentences) {
        List<String> batchTranslatedSentences = persistentTranslationService
                .translate(doc, batchSentences, doc.getFromLocale(),
                        doc.getToLocale(), backendID,
                        StringType.TEXT_PLAIN, Optional.of(CATEGORY));
        // Number of translations should match number of requests:
        assert batchSentences.size() == batchTranslatedSentences.size();
        translatedSentences.addAll(batchTranslatedSentences);
        // start a new batch
        batchSentences.clear();
    }

    /**
     * Translate strings with batch of maxLength
     * @param docStringType text/plain, text/html or text/xml
     */
    private void translateAndMergeStringsInBatch(Document doc, BackendID backendID,
            StringType docStringType, int maxLength,
            Map<Integer, TypeString> indexTypeStringMap,
            List<TypeString> results) {
        if (indexTypeStringMap.isEmpty()) return;
        List<String> batchedStrings = new ArrayList<>();
        List<Integer> indexOrderList = new ArrayList<>();
        // html or xml nodes
        Map<Integer, TranslatableNodeList> nodeCache = new HashMap<>();

        int charCount = 0;
        for (Map.Entry<Integer, TypeString> entry : indexTypeStringMap
                .entrySet()) {
            int index = entry.getKey();
            TypeString typeString = entry.getValue();

            String stringToTranslate = typeString.getValue();
            String stringType = typeString.getType();
            if (stringType.equals(MediaType.TEXT_HTML)) {
                TranslatableNodeList translatableNodeList =
                        ArticleUtil.replaceNonTranslatableNodeHTML(index,
                                stringToTranslate);
                nodeCache.put(index, translatableNodeList);
                stringToTranslate = translatableNodeList.getHtml();
            } else if (stringType.equals(MediaType.TEXT_XML)) {
                TranslatableNodeList translatableNodeList =
                        ArticleUtil.replaceNonTranslatableNodeXML(index,
                                stringToTranslate);
                nodeCache.put(index, translatableNodeList);
                stringToTranslate = translatableNodeList.getHtml();
            }

            if (charCount + stringToTranslate.length() > maxLength) {
                translateAndMergeStrings(doc, backendID,
                        StringType.fromMediaType(stringType), batchedStrings,
                        indexOrderList, results);
                charCount = 0;
                batchedStrings.clear();
                indexOrderList.clear();
            }
            batchedStrings.add(stringToTranslate);
            indexOrderList.add(entry.getKey());
            charCount += stringToTranslate.length();
        }
        translateAndMergeStrings(doc, backendID, docStringType, batchedStrings,
                indexOrderList, results);

        // restore placeholder nodes with original values in html/xml
        if (!nodeCache.isEmpty()) {
            for (Map.Entry<Integer, TranslatableNodeList> entry: nodeCache.entrySet()) {
                int index = entry.getKey();
                TypeString typeString = results.get(index);
                Map<String, Node> placeholderIdMap =
                        entry.getValue().getPlaceholderIdMap();
                String translatedString;
                switch (typeString.getType()) {
                    case MediaType.TEXT_HTML:
                        translatedString = ArticleUtil
                                .replacePlaceholderWithNode(
                                        placeholderIdMap,
                                        typeString.getValue(),
                                        ArticleUtil::wrapHTML);
                        break;
                    case MediaType.TEXT_XML:
                        translatedString = ArticleUtil
                                .replacePlaceholderWithNode(
                                        placeholderIdMap,
                                        typeString.getValue(),
                                        ArticleUtil::wrapXML);
                        break;
                    default:
                        throw new RuntimeException();
                }

                typeString.setValue(translatedString);
                results.set(index, typeString);
            }
        }
    }

    // perform translation on list of string and merge into results
    private void translateAndMergeStrings(Document doc, BackendID backendID,
            StringType stringType, List<String> strings,
            List<Integer> indexOrderList, List<TypeString> results) {

        List<String> translatedStrings = persistentTranslationService
                .translate(doc, strings, doc.getFromLocale(),
                        doc.getToLocale(), backendID, stringType,
                        Optional.of(CATEGORY));
        assert translatedStrings.size() == strings.size();

        for (int index = 0; index < translatedStrings.size(); index++) {
            results.get(indexOrderList.get(index))
                    .setValue(translatedStrings.get(index));
        }
    }

    /**
     * Translate all html tree nodes from top to bottom.
     * If parent node is being translated,
     * its child nodes will not be translated again.
     *
     * root.getAllElements().size() changes once root is being translated.
     * @param mediaType text/plain, text/html or text/xml
     */
    private StringTranslationResult translateLargeElement(Document doc, BackendID backendID,
            MediaType mediaType, int maxLength, String source,
            Function<String, Element> toElement) {

        List<APIResponse> warnings = new ArrayList<>();
        List<Node> contents =
                ArticleUtil.unwrapAsElements(toElement.apply(source));

        for (int contentIndex = 0; contentIndex < contents.size(); contentIndex++) {
            Node content = contents.get(contentIndex);
            // if content is a (large) TextNode, ie no child nodes, translate as large plain text
            if (content instanceof TextNode) {
                TextNode textNode = (TextNode) content;
                StringTranslationResult textResult =
                        translatePlainTextBySentences(doc, backendID,
                                textNode.getWholeText(), maxLength);
                textNode.text(textResult.getTranslation());
                warnings.addAll(textResult.getWarnings());
            } else {
                translateChildNodes(doc, backendID, mediaType, maxLength,
                        toElement, warnings, content);
            }
        }
        String translation = contents.stream()
                .map(Node::outerHtml)
                .collect(Collectors.joining());
        return new StringTranslationResult(translation, warnings);
    }

    private void translateChildNodes(Document doc, BackendID backendID,
            MediaType mediaType,
            int maxLength, Function<String, Element> toElement,
            List<APIResponse> warnings, Node content) {
        int childCount = content.childNodeSize();
        int childIndex = 0;

        while (childIndex < childCount) {
            Node child = content.childNode(childIndex);
            String html = child.outerHtml();
            if (html.length() <= maxLength) {
                List<String> translated =
                        persistentTranslationService
                                .translate(doc, ImmutableList.of(html),
                                        doc.getFromLocale(),
                                        doc.getToLocale(), backendID,
                                        StringType.fromMediaType(mediaType),
                                        Optional.of(CATEGORY));
                assert translated.size() == 1;
                Node replacement = ArticleUtil
                        .asElement(translated.get(0), toElement);
                if (replacement != null) {
                    child.replaceWith(replacement);
                }
            } else {
                // if child is a (large) TextNode, ie no child nodes, translate as large plain text
                if (child instanceof TextNode) {
                    TextNode textNode = (TextNode) child;
                    StringTranslationResult textResult =
                            translatePlainTextBySentences(doc, backendID,
                                    textNode.getWholeText(), maxLength);
                    textNode.text(textResult.getTranslation());
                    warnings.addAll(textResult.getWarnings());
                } else {
                    // show warning if there are no more children under this node
                    warnings.add(maxLengthWarning(html, maxLength));
                }
            }
            // size changes if child node is being translated
            childCount = content.childNodeSize();
            childIndex++;
        }
    }

    private APIResponse maxLengthWarning(String source,
            int maxLength) {
        String title =
                "Warning: translation skipped: String length is over " +
                        maxLength;
        String shortenString = ShortString.shorten(source);
        LOG.warn(title + " - " + shortenString);
        return new APIResponse(Response.Status.BAD_REQUEST,
                new Exception(shortenString), title);
    }

    public MediaType getMediaType(String mediaType) throws BadRequestException {
        if (isMediaTypeSupported(mediaType)) {
            return MediaType.valueOf(mediaType);
        }
        throw new BadRequestException("Unsupported media type: " + mediaType);
    }

    public boolean isMediaTypeSupported(String mediaType) {
        return StringUtils.equalsAny(mediaType, MediaType.TEXT_HTML,
                MediaType.TEXT_PLAIN, MediaType.TEXT_XML);
    }
}

File: business-svc/src/main/java/org/zanata/magpie/service/DocumentContentTranslatorService.kt
package org.zanata.magpie.service

import java.util.ArrayList
import java.util.HashMap
import java.util.LinkedHashMap
import java.util.Optional
import java.util.stream.Collectors
import javax.ejb.Stateless
import javax.inject.Inject
import javax.ws.rs.BadRequestException
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response

import org.apache.commons.lang3.StringUtils
import org.jsoup.nodes.Element
import org.slf4j.LoggerFactory
import org.zanata.magpie.api.dto.APIResponse
import org.zanata.magpie.api.dto.DocumentContent
import org.zanata.magpie.api.dto.TypeString
import org.zanata.magpie.model.Document
import org.zanata.magpie.model.StringType
import org.zanata.magpie.util.ArticleUtil
import org.zanata.magpie.exception.MTException
import org.zanata.magpie.model.BackendID

import com.google.common.collect.ImmutableList
import org.zanata.magpie.api.dto.LocaleCode
import org.zanata.magpie.util.SegmentString
import org.zanata.magpie.util.ShortString

/**
 * @author Alex Eng [aeng@redhat.com](mailto:aeng@redhat.com)
 */
@Stateless
class DocumentContentTranslatorService @Inject constructor(
        private val persistentTranslationService: PersistentTranslationService) {

    /**
     * Translate a Document and send in for machine translation request.
     *
     * If string is no longer than maxLength, it will be send as a batch according
     * to the type.
     *
     * If a string is longer than maxLength,
     * Plain Text: try to segment it and send it as a single MT request. Will
     * be ignored if cannot be segmented.
     * HTML: try to run through the node tree and translate. Will be ignored if
     * cannot be broken down.
     *
     * [DocumentContent]
     */
    @Throws(BadRequestException::class, MTException::class)
    fun translateDocument(doc: Document, documentContent: DocumentContent,
            backendID: BackendID): DocumentContent {
        val indexTextMap = LinkedHashMap<Int, TypeString>()
        val indexHTMLMap = LinkedHashMap<Int, TypeString>()
        val indexXMLMap = LinkedHashMap<Int, TypeString>()
        val maxLength = persistentTranslationService.getMaxLength(backendID)

        val warnings = ArrayList<APIResponse>()
        val typeStrings = ArrayList(documentContent.contents)

        // counts short strings (less than maxLength) which are translated
        // in a batch at the end
        for ((index, typeString) in typeStrings.withIndex()) {
            val mediaType = getMediaType(typeString.type)
            val source = typeString.value
            if (mediaType == MediaType.TEXT_PLAIN_TYPE) {
                if (source.length <= maxLength) {
                    indexTextMap[index] = typeString
                } else {
                    val (translation, warnings1) = translateLargeString(doc, backendID,
                            StringType.TEXT_PLAIN, source, maxLength)
                    typeString.value = translation
                    warnings.addAll(warnings1)
                }
            } else if (mediaType == MediaType.TEXT_HTML_TYPE || mediaType == MediaType.TEXT_XML_TYPE) {
                val xml = mediaType == MediaType.TEXT_XML_TYPE

                // put placeholders in all non-translatable elements
                val translatableNodeList = if (xml)
                    ArticleUtil.replaceNonTranslatableNodeXML(index, source)
                else
                    ArticleUtil.replaceNonTranslatableNodeHTML(index, source)
                val html = translatableNodeList.html
                if (html.length <= maxLength) {
                    val indexMap = if (xml) indexXMLMap else indexHTMLMap
                    indexMap[index] = typeString
                } else {
                    val toElement = if (xml) ArticleUtil::wrapXML else ArticleUtil::wrapHTML
                    val (translationWithPlaceholders, warnings1) = translateLargeElement(doc, backendID, mediaType,
                            maxLength, html, toElement)

                    // replace placeholder with original node
                    val translatedString = removePlaceholders(translationWithPlaceholders,
                            translatableNodeList, toElement)
                    typeString.value = translatedString
                    warnings.addAll(warnings1)
                }
            }
        }

        translateAndMergeStringsInBatch(doc, backendID, StringType.TEXT_PLAIN,
                maxLength, indexTextMap, typeStrings)
        translateAndMergeStringsInBatch(doc, backendID, StringType.HTML,
                maxLength, indexHTMLMap, typeStrings)
        translateAndMergeStringsInBatch(doc, backendID, StringType.XML,
                maxLength, indexXMLMap, typeStrings)
        return DocumentContent(typeStrings, documentContent.url,
                doc.toLocale.localeCode.id, backendID.id,
                warnings)
    }

    private fun removePlaceholders(translationWithPlaceholders: String,
            translatableNodeList: TranslatableNodeList,
            toElement: (String) -> Element): String {
        return ArticleUtil.replacePlaceholderWithNode(
                translatableNodeList.placeholderIdMap,
                translationWithPlaceholders, toElement)
    }

    /**
     * Translate strings with batch of maxLength
     */
    private fun translateLargeString(doc: Document, backendID: BackendID,
            stringType: StringType, source: String, maxLength: Int): StringTranslationResult {
        val warnings = ArrayList<APIResponse>()
        val segmentedStrings = SegmentString.segmentString(source,
                Optional.of<LocaleCode>(doc.fromLocale.localeCode))
        val results = ArrayList(segmentedStrings)

        val batchedStrings = ArrayList<String>()
        val indexOrderList = ArrayList<Int>()
        val translatedStrings = ArrayList<String>()
        var charCount = 0
        for (index in segmentedStrings.indices) {
            val string = segmentedStrings[index]
            // ignore string if length is longer the maxLength
            if (string.length > maxLength) {
                warnings.add(maxLengthWarning(string, maxLength))
                continue
            }
            if (charCount + string.length > maxLength) {
                val translated = persistentTranslationService
                        .translate(doc, batchedStrings, doc.fromLocale,
                                doc.toLocale, backendID, stringType,
                                Optional.of(CATEGORY))
                translatedStrings.addAll(translated)
                assert(batchedStrings.size == translated.size)
                charCount = 0
                batchedStrings.clear()
            }
            batchedStrings.add(string)
            indexOrderList.add(index)
            charCount += string.length
        }
        if (!batchedStrings.isEmpty()) {
            val translated = persistentTranslationService
                    .translate(doc, batchedStrings, doc.fromLocale,
                            doc.toLocale, backendID, stringType,
                            Optional.of(CATEGORY))
            translatedStrings.addAll(translated)
            assert(batchedStrings.size == translated.size)
        }

        for (index in translatedStrings.indices) {
            results[indexOrderList[index]] = translatedStrings[index]
        }
        return StringTranslationResult(results.joinToString(""), warnings)
    }

    /**
     * Translate strings with batch of maxLength
     * @param docStringType text/plain, text/html or text/xml
     */
    private fun translateAndMergeStringsInBatch(doc: Document, backendID: BackendID,
            docStringType: StringType, maxLength: Int,
            indexTypeStringMap: Map<Int, TypeString>,
            results: MutableList<TypeString>) {
        if (indexTypeStringMap.isEmpty()) return
        val batchedStrings = ArrayList<String>()
        val indexOrderList = ArrayList<Int>()
        // html or xml nodes
        val nodeCache = HashMap<Int, TranslatableNodeList>()

        var charCount = 0
        for ((index, typeString) in indexTypeStringMap) {

            var stringToTranslate = typeString.value
            val stringType = typeString.type
            if (stringType == MediaType.TEXT_HTML) {
                val translatableNodeList = ArticleUtil.replaceNonTranslatableNodeHTML(index,
                        stringToTranslate)
                nodeCache[index] = translatableNodeList
                stringToTranslate = translatableNodeList.html
            } else if (stringType == MediaType.TEXT_XML) {
                val translatableNodeList = ArticleUtil.replaceNonTranslatableNodeXML(index,
                        stringToTranslate)
                nodeCache[index] = translatableNodeList
                stringToTranslate = translatableNodeList.html
            }

            if (charCount + stringToTranslate.length > maxLength) {
                translateAndMergeStrings(doc, backendID,
                        StringType.fromMediaType(stringType), batchedStrings,
                        indexOrderList, results)
                charCount = 0
                batchedStrings.clear()
                indexOrderList.clear()
            }
            batchedStrings.add(stringToTranslate)
            indexOrderList.add(index)
            charCount += stringToTranslate.length
        }
        translateAndMergeStrings(doc, backendID, docStringType, batchedStrings,
                indexOrderList, results)

        // restore placeholder nodes with original values in html/xml
        for ((index, value) in nodeCache) {
            val typeString = results[index]
            val placeholderIdMap = value.placeholderIdMap
            val translatedString = when (typeString.type) {
                MediaType.TEXT_HTML -> ArticleUtil.replacePlaceholderWithNode(
                        placeholderIdMap, typeString.value, ArticleUtil::wrapHTML
                )
                MediaType.TEXT_XML -> ArticleUtil.replacePlaceholderWithNode(
                        placeholderIdMap, typeString.value, ArticleUtil::wrapXML
                )
                else -> throw RuntimeException()
            }
            typeString.value = translatedString
            results[index] = typeString
        }
    }

    // perform translation on list of string and merge into results
    private fun translateAndMergeStrings(doc: Document, backendID: BackendID,
            stringType: StringType, strings: List<String>,
            indexOrderList: List<Int>, results: List<TypeString>) {
        val translatedStrings = persistentTranslationService
                .translate(doc, strings, doc.fromLocale,
                        doc.toLocale, backendID, stringType,
                        Optional.of(CATEGORY))
        assert(translatedStrings.size == strings.size)

        for (index in translatedStrings.indices) {
            results[indexOrderList[index]].value = translatedStrings[index]
        }
    }

    /**
     * Translate all html tree nodes from top to bottom.
     * If parent node is being translated,
     * its child nodes will not be translated again.
     *
     * root.getAllElements().size() changes once root is being translated.
     * @param mediaType text/plain, text/html or text/xml
     */
    private fun translateLargeElement(doc: Document, backendID: BackendID,
            mediaType: MediaType, maxLength: Int, source: String,
            toElement: (String) -> Element): StringTranslationResult {

        val warnings = ArrayList<APIResponse>()
        val contents = ArticleUtil.unwrapAsElements(toElement(source))

        // TODO try handling as Nodes, using childNode() and childNodeSize()
        for (i in contents.indices) {
            val content = contents[i]
            var size = content.allElements.size
            var index = 0

            while (index < size) {
                val child = content.allElements[index]
                val html = child.outerHtml()
                if (html.length <= maxLength) {
                    val translated = persistentTranslationService
                            .translate(doc, ImmutableList.of(html),
                                    doc.fromLocale,
                                    doc.toLocale, backendID,
                                    StringType.fromMediaType(mediaType),
                                    Optional.of(CATEGORY))
                    assert(translated.size == 1)
                    val replacement = ArticleUtil.asElement(translated[0], toElement)
                    if (replacement != null) {
                        if (child === content) {
                            //replace this item in contents list, exit while loop
                            contents[i] = replacement
                            break
                        } else {
                            child.replaceWith(replacement)
                        }
                    }
                } else {
                    // show warning if there are no more children under this node
                    warnings.add(maxLengthWarning(html, maxLength))
                }
                // size changes if child node is being translated
                size = content.allElements.size
                index++
            }
        }
        val translation = contents.joinToString("", transform = Element::outerHtml)
        return StringTranslationResult(translation, warnings)
    }

    private fun maxLengthWarning(source: String,
            maxLength: Int): APIResponse {
        val title = "Warning: translation skipped: String length is over $maxLength"
        val shortenString = ShortString.shorten(source)
        LOG.warn("$title - $shortenString")
        return APIResponse(Response.Status.BAD_REQUEST,
                Exception(shortenString), title)
    }

    @Throws(BadRequestException::class)
    fun getMediaType(mediaType: String): MediaType {
        if (isMediaTypeSupported(mediaType)) {
            return MediaType.valueOf(mediaType)
        }
        throw BadRequestException("Unsupported media type: $mediaType")
    }

    fun isMediaTypeSupported(mediaType: String): Boolean {
        return StringUtils.equalsAny(mediaType, MediaType.TEXT_HTML,
                MediaType.TEXT_PLAIN, MediaType.TEXT_XML)
    }

    companion object {
        private val LOG = LoggerFactory.getLogger(DocumentContentTranslatorService::class.java)

        // set MT category for translation
        private const val CATEGORY = "tech"
    }
}
